File: samples/typescript_nodejs/13.core-bot/src/dialogs/luisHelper.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { BookingDetails } from './bookingDetails';

import { RecognizerResult, TurnContext } from 'botbuilder';

import { LuisRecognizer } from 'botbuilder-ai';

export class LuisHelper {
    /**
     * Returns an object with preformatted LUIS results for the bot's dialogs to consume.
     * @param {TurnContext} context
     */
    public static async executeLuisQuery(context: TurnContext): Promise<BookingDetails> {
        const bookingDetails = new BookingDetails();

        try {
            const recognizer = new LuisRecognizer({
                applicationId: process.env.LuisAppId,
                endpoint: `https://${ process.env.LuisAPIHostName }`,
                endpointKey: process.env.LuisAPIKey,
            }, {}, true);

            const recognizerResult = await recognizer.recognize(context);

            const intent = LuisRecognizer.topIntent(recognizerResult);

            bookingDetails.intent = intent;

            if (intent === 'Book_flight') {
                // We need to get the result from the LUIS JSON which at every level returns an array

                bookingDetails.destination = LuisHelper.parseCompositeEntity(recognizerResult, 'To', 'Airport');
                bookingDetails.origin = LuisHelper.parseCompositeEntity(recognizerResult, 'From', 'Airport');

                // This value will be a TIMEX. And we are only interested in a Date so grab the first result and drop the Time part.
                // TIMEX is a format that represents DateTime expressions that include some ambiguity. e.g. missing a Year.
                bookingDetails.travelDate = LuisHelper.parseDatetimeEntity(recognizerResult);
            }
        } catch (err) {
            console.warn(`LUIS Exception: ${ err } Check your LUIS configuration`);
        }
        return bookingDetails;
    }

    private static parseCompositeEntity(result: RecognizerResult, compositeName: string, entityName: string): string {
        const compositeEntity = result.entities[compositeName];
        if (!compositeEntity || !compositeEntity[0]) {
            return undefined;
        }

        const entity = compositeEntity[0][entityName];
        if (!entity || !entity[0]) {
            return undefined;
        }

        const entityValue = entity[0][0];
        return entityValue;
    }

    private static parseDatetimeEntity(result: RecognizerResult): string {
        const datetimeEntity = result.entities.datetime;
        if (!datetimeEntity || !datetimeEntity[0]) {
            return undefined;
        }

        const timex = datetimeEntity[0].timex;
        if (!timex || !timex[0]) {
            return undefined;
        }

        const datetime = timex[0].split('T')[0];
        return datetime;
    }
}

File: samples/python/21.corebot-app-insights/helpers/luis_helper.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

"""Helper to call LUIS service."""
from botbuilder.ai.luis import LuisRecognizer, LuisApplication, LuisPredictionOptions
from botbuilder.core import TurnContext, BotTelemetryClient, NullTelemetryClient

from booking_details import BookingDetails

# pylint: disable=line-too-long
class LuisHelper:
    """LUIS helper implementation."""

    @staticmethod
    async def execute_luis_query(
        configuration,
        turn_context: TurnContext,
        telemetry_client: BotTelemetryClient = None,
    ) -> BookingDetails:
        """Invoke LUIS service to perform prediction/evaluation of utterance."""
        booking_details = BookingDetails()

        # pylint:disable=broad-except
        try:
            luis_application = LuisApplication(
                configuration.get("LUIS_APP_ID"),
                configuration.get("LUIS_API_KEY"),
                configuration.get("LUIS_API_HOST_NAME"),
            )
            options = LuisPredictionOptions()
            options.telemetry_client = (
                telemetry_client
                if telemetry_client is not None
                else NullTelemetryClient()
            )
            recognizer = LuisRecognizer(luis_application, prediction_options=options)
            recognizer_result = await recognizer.recognize(turn_context)
            print(f"Recognize Result: {recognizer_result}")

            if recognizer_result.intents:
                intent = sorted(
                    recognizer_result.intents,
                    key=recognizer_result.intents.get,
                    reverse=True,
                )[:1][0]
                if intent == "Book_flight":
                    # We need to get the result from the LUIS JSON which at every level returns an array.
                    to_entities = recognizer_result.entities.get("$instance", {}).get(
                        "To", []
                    )
                    if to_entities:
                        booking_details.destination = to_entities[0]["text"]
                    from_entities = recognizer_result.entities.get("$instance", {}).get(
                        "From", []
                    )
                    if from_entities:
                        booking_details.origin = from_entities[0]["text"]

                    # This value will be a TIMEX. And we are only interested in a Date so grab the first result and drop the Time part.
                    # TIMEX is a format that represents DateTime expressions that include some ambiguity. e.g. missing a Year.
                    date_entities = recognizer_result.entities.get("$instance", {}).get(
                        "datetime", []
                    )
                    if date_entities:
                        booking_details.travel_date = (
                            None  # Set when we get a timex format
                        )
        except Exception as exception:
            print(exception)

        return booking_details
